tests/Compiler/MapperFactory/DefaultMapperCompilerFactoryTest.php

```diff
@@ -0,0 +1,350 @@
+use ShipMonk\InputMapper\Compiler\Mapper\Wrapper\ValidatedMapperCompiler;
+use ShipMonk\InputMapper\Compiler\MapperFactory\DefaultMapperCompilerFactory;
+use ShipMonk\InputMapper\Compiler\Validator\Int\AssertIntRange;
+use ShipMonk\InputMapper\Compiler\Validator\Int\AssertPositiveInt;
+use ShipMonk\InputMapper\Compiler\Validator\String\AssertStringLength;
+use ShipMonkTests\InputMapper\Compiler\MapperFactory\Data\BrandInput;
+use ShipMonkTests\InputMapper\Compiler\MapperFactory\Data\CarInput;
+use ShipMonkTests\InputMapper\Compiler\MapperFactory\Data\ColorEnum;
+use ShipMonkTests\InputMapper\Compiler\MapperFactory\Data\InputWithDate;
+use ShipMonkTests\InputMapper\Compiler\MapperFactory\Data\InputWithoutConstructor;
+use ShipMonkTests\InputMapper\Compiler\MapperFactory\Data\InputWithPrivateConstructor;
+use ShipMonkTests\InputMapper\InputMapperTestCase;
+
+class DefaultMapperCompilerFactoryTest extends InputMapperTestCase
+{
+
+    /**
+     * @param  array<string, mixed> $options
+     */
+    #[DataProvider('provideCreateOkData')]
+    public function testCreateOk(string $type, array $options, MapperCompiler $expectedMapperCompiler): void
+    {
+        $phpDocLexer = new Lexer();
+        $phpDocExprParser = new ConstExprParser(unescapeStrings: true);
+        $phpDocTypeParser = new TypeParser($phpDocExprParser);
+        $phpDocParser = new PhpDocParser($phpDocTypeParser, $phpDocExprParser);
+        $phpDocType = $phpDocTypeParser->parse(new TokenIterator($phpDocLexer->tokenize($type)));
+
+        $mapperCompilerFactory = new DefaultMapperCompilerFactory($phpDocLexer, $phpDocParser);
+        $mapperCompiler = $mapperCompilerFactory->create($phpDocType, $options);
+
+        self::assertEquals($expectedMapperCompiler, $mapperCompiler);
+    }
+
+    /**
+     * @return iterable<array{string, array<string, mixed>, MapperCompiler}>
+     */
+    public static function provideCreateOkData(): iterable
+    {
+        yield 'CarInput' => [
+            CarInput::class,
+            [],
+            new MapObject(CarInput::class, [
+                'id' => new MapInt(),
+                'name' => new ValidatedMapperCompiler(new MapString(), [new AssertStringLength(exact: 7)]),
+                'brand' => new MapOptional(new DelegateMapperCompiler(BrandInput::class)),
+                'numbers' => new MapList(new ValidatedMapperCompiler(new MapInt(), [new AssertPositiveInt()])),
+            ]),
+        ];
+
+        yield 'CarInput with forced root delegation' => [
+            CarInput::class,
+            [DefaultMapperCompilerFactory::DELEGATE_OBJECT_MAPPING => true],
+            new DelegateMapperCompiler(CarInput::class),
+        ];
+
+        yield 'BrandInput (with allowed extra keys)' => [
+            BrandInput::class,
+            [],
+            new MapObject(
+                BrandInput::class,
+                ['name' => new MapString()],
+                allowExtraProperties: true,
```

Comment:
We should unify naming: allowExtraProperties vs AllowExtraKeys. Or maybe allowExtraFields. 

https://github.com/brick/json-mapper uses "extra properties".